The current HPX master branch is still not compatible with Kokkos 4.0.1

## Expected Behavior

The current HPX master branch (cf70bf1f) can be built with Kokkos 4.0.1.

## Actual Behavior

I got a bunch of errors such as

```
3195    /global/u1/j/jackyan/workspace/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/hpx-master-2yhwuuses2ylidcqlecleitp7ubqotmi/include/hpx/parallel/util/detail/chunk_size_iterator.hpp(350): warning #20012-D: __host__ annotation is ignored on a function("chunk_size_idx_iterator") t
             hat is explicitly defaulted on its first declaration
     3196    
     3197    /global/u1/j/jackyan/workspace/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/hpx-master-2yhwuuses2ylidcqlecleitp7ubqotmi/include/hpx/parallel/util/detail/chunk_size_iterator.hpp(350): warning #20012-D: __device__ annotation is ignored on a function("chunk_size_idx_iterator")
              that is explicitly defaulted on its first declaration
     3198    
     3199    /global/u1/j/jackyan/workspace/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/hpx-master-2yhwuuses2ylidcqlecleitp7ubqotmi/include/hpx/functional/experimental/scope_exit.hpp: In instantiation of 'struct hpx::experimental::scope_exit<hpx::threads::coroutines::stackless_coroutin
             e::operator()(hpx::threads::coroutines::stackless_coroutine::arg_type)::__T0>':
     3200    /global/u1/j/jackyan/workspace/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/hpx-master-2yhwuuses2ylidcqlecleitp7ubqotmi/include/hpx/coroutines/stackless_coroutine.hpp:281:48:   required from here
  >> 3201    /global/u1/j/jackyan/workspace/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/hpx-master-2yhwuuses2ylidcqlecleitp7ubqotmi/include/hpx/functional/experimental/scope_exit.hpp:63:12: error: 'hpx::experimental::scope_exit<F>::f' has incomplete type
     3202       63 |         F f;
     3203          |            ^
     3204    /global/u1/j/jackyan/workspace/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/hpx-master-2yhwuuses2ylidcqlecleitp7ubqotmi/include/hpx/coroutines/stackless_coroutine.hpp:280:50: note: forward declaration of 'class hpx::threads::coroutines::stackless_coroutine::operator()(hpx::
             threads::coroutines::stackless_coroutine::arg_type)::__T0'
     3205      280 |             auto on_exit = hpx::experimental::scope_exit(
     3206          |                                                  ^~~~
     3207    /global/u1/j/jackyan/workspace/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/hpx-master-2yhwuuses2ylidcqlecleitp7ubqotmi/include/hpx/coroutines/stackless_coroutine.hpp: In member function 'hpx::threads::coroutines::stackless_coroutine::result_type hpx::threads::coroutines::s
             tackless_coroutine::operator()(hpx::threads::coroutines::stackless_coroutine::arg_type)':
  >> 3208    /global/u1/j/jackyan/workspace/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/hpx-master-2yhwuuses2ylidcqlecleitp7ubqotmi/include/hpx/coroutines/stackless_coroutine.hpp:281:48: error: no matching function for call to 'hpx::experimental::scope_exit<hpx::threads::coroutines::st
             ackless_coroutine::operator()(hpx::threads::coroutines::stackless_coroutine::arg_type)::__T0>::scope_exit(hpx::threads::coroutines::stackless_coroutine::operator()(hpx::threads::coroutines::stackless_coroutine::arg_type)::<lambda()>)'
     3209      281 |                 [] { detail::coroutine_self::set_self(nullptr); });
     3210          |                                                ^
     3211    /global/u1/j/jackyan/workspace/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/hpx-master-2yhwuuses2ylidcqlecleitp7ubqotmi/include/hpx/functional/experimental/scope_exit.hpp:36:11: note: candidate: 'constexpr hpx::experimental::scope_exit<F>::scope_exit(hpx::experimental::scop
             e_exit<F>&&) [with F = hpx::threads::coroutines::stackless_coroutine::operator()(hpx::threads::coroutines::stackless_coroutine::arg_type)::__T0]'
     3212       36 |         constexpr scope_exit(scope_exit&& rhs) noexcept(
     3213          |           ^~~~~~~~~~
     3214    /global/u1/j/jackyan/workspace/spack/opt/spack/linux-sles15-zen3/gcc-11.2.0/hpx-master-2yhwuuses2ylidcqlecleitp7ubqotmi/include/hpx/functional/experimental/scope_exit.hpp:36:35: note:   no known conversion for argument 1 from 'hpx::threads::coroutines::stackless_coroutine::
             operator()(hpx::threads::coroutines::stackless_coroutine::arg_type)::<lambda()>' to 'hpx::experimental::scope_exit<hpx::threads::coroutines::stackless_coroutine::operator()(hpx::threads::coroutines::stackless_coroutine::arg_type)::__T0>&&'
```

## Specifications

It has been reproduced on Perlmutter and Polaris.

The full spack build log: [spack-build-out.txt](https://github.com/STEllAR-GROUP/hpx/files/13935006/spack-build-out.txt)

cc @G-071

--- a/libs/core/functional/include/hpx/functional/experimental/scope_exit.hpp
+++ b/libs/core/functional/include/hpx/functional/experimental/scope_exit.hpp
@@ -17,55 +17,61 @@
 
 namespace hpx::experimental {
 
-    template <typename F>
-    struct scope_exit
-    {
-        explicit constexpr scope_exit(F&& f) noexcept(
-            std::is_nothrow_move_constructible_v<F> ||
-            std::is_nothrow_copy_constructible_v<F>)
-          : f(HPX_MOVE(f))
-        {
-        }
+    namespace detail {
 
-        explicit constexpr scope_exit(F const& f) noexcept(
-            std::is_nothrow_copy_constructible_v<F>)
-          : f(f)
+        template <typename F>
+        struct scope_exit
         {
-        }
-
-        constexpr scope_exit(scope_exit&& rhs) noexcept(
-            std::is_nothrow_move_constructible_v<F> ||
-            std::is_nothrow_copy_constructible_v<F>)
-          : f(HPX_MOVE(rhs.f))
-          , active(rhs.active)
-        {
-            rhs.release();
-        }
-
-        scope_exit(scope_exit const&) = delete;
-        scope_exit& operator=(scope_exit const&) = delete;
-        scope_exit& operator=(scope_exit&& rhs) = delete;
-
-        HPX_CONSTEXPR_DESTRUCTOR ~scope_exit() noexcept
-        {
-            if (active)
+            explicit constexpr scope_exit(F&& f) noexcept(
+                std::is_nothrow_move_constructible_v<F> ||
+                std::is_nothrow_copy_constructible_v<F>)
+              : f(HPX_MOVE(f))
             {
-                f();
             }
-        }
 
-        constexpr void release() noexcept
-        {
-            active = false;
-        }
+            explicit constexpr scope_exit(F const& f) noexcept(
+                std::is_nothrow_copy_constructible_v<F>)
+              : f(f)
+            {
+            }
 
-    private:
-        F f;
-        bool active = true;
-    };
+            constexpr scope_exit(scope_exit&& rhs) noexcept(
+                std::is_nothrow_move_constructible_v<F> ||
+                std::is_nothrow_copy_constructible_v<F>)
+              : f(HPX_MOVE(rhs.f))
+              , active(rhs.active)
+            {
+                rhs.release();
+            }
+
+            scope_exit(scope_exit const&) = delete;
+            scope_exit& operator=(scope_exit const&) = delete;
+            scope_exit& operator=(scope_exit&& rhs) = delete;
+
+            HPX_CONSTEXPR_DESTRUCTOR ~scope_exit() noexcept
+            {
+                if (active)
+                {
+                    f();
+                }
+            }
+
+            constexpr void release() noexcept
+            {
+                active = false;
+            }
+
+        private:
+            F f;
+            bool active = true;
+        };
+    }    // namespace detail
 
     template <typename F>
-    scope_exit(F) -> scope_exit<F>;
+    auto scope_exit(F&& f)
+    {
+        return detail::scope_exit<std::decay_t<F>>(HPX_FORWARD(F, f));
+    }
 }    // namespace hpx::experimental
 
 #endif

--- a/libs/core/functional/include/hpx/functional/experimental/scope_fail.hpp
+++ b/libs/core/functional/include/hpx/functional/experimental/scope_fail.hpp
@@ -19,57 +19,63 @@
 
 namespace hpx::experimental {
 
-    template <typename F>
-    struct scope_fail
-    {
-        explicit constexpr scope_fail(F&& f) noexcept(
-            std::is_nothrow_move_constructible_v<F> ||
-            std::is_nothrow_copy_constructible_v<F>)
-          : f(HPX_MOVE(f))
-          , active(std::uncaught_exceptions())
-        {
-        }
+    namespace detail {
 
-        explicit constexpr scope_fail(F const& f) noexcept(
-            std::is_nothrow_copy_constructible_v<F>)
-          : f(f)
-          , active(std::uncaught_exceptions())
+        template <typename F>
+        struct scope_fail
         {
-        }
-
-        constexpr scope_fail(scope_fail&& rhs) noexcept(
-            std::is_nothrow_move_constructible_v<F> ||
-            std::is_nothrow_copy_constructible_v<F>)
-          : f(HPX_MOVE(rhs.f))
-          , active(rhs.active)
-        {
-            rhs.release();
-        }
-
-        scope_fail(scope_fail const&) = delete;
-        scope_fail& operator=(scope_fail const&) = delete;
-        scope_fail& operator=(scope_fail&& rhs) = delete;
-
-        HPX_CONSTEXPR_DESTRUCTOR ~scope_fail() noexcept
-        {
-            if (active < std::uncaught_exceptions())
+            explicit constexpr scope_fail(F&& f) noexcept(
+                std::is_nothrow_move_constructible_v<F> ||
+                std::is_nothrow_copy_constructible_v<F>)
+              : f(HPX_MOVE(f))
+              , active(std::uncaught_exceptions())
             {
-                f();
             }
-        }
 
-        constexpr void release() noexcept
-        {
-            active = (std::numeric_limits<int>::max)();
-        }
+            explicit constexpr scope_fail(F const& f) noexcept(
+                std::is_nothrow_copy_constructible_v<F>)
+              : f(f)
+              , active(std::uncaught_exceptions())
+            {
+            }
 
-    private:
-        F f;
-        int active;
-    };
+            constexpr scope_fail(scope_fail&& rhs) noexcept(
+                std::is_nothrow_move_constructible_v<F> ||
+                std::is_nothrow_copy_constructible_v<F>)
+              : f(HPX_MOVE(rhs.f))
+              , active(rhs.active)
+            {
+                rhs.release();
+            }
+
+            scope_fail(scope_fail const&) = delete;
+            scope_fail& operator=(scope_fail const&) = delete;
+            scope_fail& operator=(scope_fail&& rhs) = delete;
+
+            HPX_CONSTEXPR_DESTRUCTOR ~scope_fail() noexcept
+            {
+                if (active < std::uncaught_exceptions())
+                {
+                    f();
+                }
+            }
+
+            constexpr void release() noexcept
+            {
+                active = (std::numeric_limits<int>::max)();
+            }
+
+        private:
+            F f;
+            int active;
+        };
+    }    // namespace detail
 
     template <typename F>
-    scope_fail(F) -> scope_fail<F>;
+    auto scope_fail(F&& f)
+    {
+        return detail::scope_fail<std::decay_t<F>>(HPX_FORWARD(F, f));
+    }
 }    // namespace hpx::experimental
 
 #endif

--- a/libs/core/functional/include/hpx/functional/experimental/scope_success.hpp
+++ b/libs/core/functional/include/hpx/functional/experimental/scope_success.hpp
@@ -18,57 +18,64 @@
 
 namespace hpx::experimental {
 
-    template <typename F>
-    struct scope_success
-    {
-        explicit constexpr scope_success(F&& f) noexcept(
-            std::is_nothrow_move_constructible_v<F> ||
-            std::is_nothrow_copy_constructible_v<F>)
-          : f(HPX_MOVE(f))
-          , active(std::uncaught_exceptions())
-        {
-        }
+    namespace detail {
 
-        explicit constexpr scope_success(F const& f) noexcept(
-            std::is_nothrow_copy_constructible_v<F>)
-          : f(f)
-          , active(std::uncaught_exceptions())
+        template <typename F>
+        struct scope_success
         {
-        }
-
-        constexpr scope_success(scope_success&& rhs) noexcept(
-            std::is_nothrow_move_constructible_v<F> ||
-            std::is_nothrow_copy_constructible_v<F>)
-          : f(HPX_MOVE(rhs.f))
-          , active(rhs.active)
-        {
-            rhs.release();
-        }
-
-        scope_success(scope_success const&) = delete;
-        scope_success& operator=(scope_success const&) = delete;
-        scope_success& operator=(scope_success&& rhs) = delete;
-
-        HPX_CONSTEXPR_DESTRUCTOR ~scope_success() noexcept(noexcept(this->f()))
-        {
-            if (active >= std::uncaught_exceptions())
+            explicit constexpr scope_success(F&& f) noexcept(
+                std::is_nothrow_move_constructible_v<F> ||
+                std::is_nothrow_copy_constructible_v<F>)
+              : f(HPX_MOVE(f))
+              , active(std::uncaught_exceptions())
             {
-                f();
             }
-        }
 
-        constexpr void release() noexcept
-        {
-            active = -1;
-        }
+            explicit constexpr scope_success(F const& f) noexcept(
+                std::is_nothrow_copy_constructible_v<F>)
+              : f(f)
+              , active(std::uncaught_exceptions())
+            {
+            }
 
-    private:
-        F f;
-        int active;
-    };
+            constexpr scope_success(scope_success&& rhs) noexcept(
+                std::is_nothrow_move_constructible_v<F> ||
+                std::is_nothrow_copy_constructible_v<F>)
+              : f(HPX_MOVE(rhs.f))
+              , active(rhs.active)
+            {
+                rhs.release();
+            }
+
+            scope_success(scope_success const&) = delete;
+            scope_success& operator=(scope_success const&) = delete;
+            scope_success& operator=(scope_success&& rhs) = delete;
+
+            HPX_CONSTEXPR_DESTRUCTOR ~scope_success() noexcept(
+                noexcept(this->f()))
+            {
+                if (active >= std::uncaught_exceptions())
+                {
+                    f();
+                }
+            }
+
+            constexpr void release() noexcept
+            {
+                active = -1;
+            }
+
+        private:
+            F f;
+            int active;
+        };
+    }    // namespace detail
 
     template <typename F>
-    scope_success(F) -> scope_success<F>;
+    auto scope_success(F&& f)
+    {
+        return detail::scope_success<std::decay_t<F>>(HPX_FORWARD(F, f));
+    }
 }    // namespace hpx::experimental
 
 #endif

